src/selectors/AutoTuner.h

```diff
@@ -0,0 +1,126 @@
+   * @param verletRebuildFrequency Specifies after how many pair-wise traversals the neighbor lists are to be rebuild.
+   * @param allowedContainerOptions Vector of container types AutoPas can choose from.
+   * @param allowedTraversalOptions Vector of traversals AutoPas can choose from.
+   * @param tuningInterval Number of timesteps after which the auto-tuner shall reevaluate all selections.
+   */
+  AutoTuner(std::array<double, 3> boxMin, std::array<double, 3> boxMax, double cutoff, double verletSkin,
+            unsigned int verletRebuildFrequency, std::vector<ContainerOptions> allowedContainerOptions,
+            std::vector<TraversalOptions> allowedTraversalOptions, unsigned int tuningInterval)
+      : _tuningInterval(tuningInterval),
+        _iterationsSinceTuning(tuningInterval),  // init to max so that tuning happens in first iteration
+        _containerSelector(boxMin, boxMax, cutoff, verletSkin, verletRebuildFrequency, allowedContainerOptions,
+                           allowedTraversalOptions) {}
+
+  /**
+   * Getter for the optimal container.
+   * @return Smartpointer to the optimal container.
+   */
+  std::shared_ptr<autopas::ParticleContainer<Particle, ParticleCell>> getContainer() {
+    return _containerSelector.getOptimalContainer();
+  };
+
+  /**
+   * Function to iterate over all pairs of particles in the container.
+   * This function only handles short-range interactions.
+   * @tparam ParticleFunctor
+   * @param f Functor that describes the pair-potential
+   * @param useSoA if true SoA data structure is used otherwise AoS
+   */
+  template <class ParticleFunctor>
+  void iteratePairwise(ParticleFunctor *f, bool useSoA);
+
+ private:
+  template <class PairwiseFunctor, bool useSoA, bool useNewton3>
+  void tune(PairwiseFunctor &pairwiseFunctor);
+
+  unsigned int _tuningInterval, _iterationsSinceTuning;
+  ContainerSelector<Particle, ParticleCell> _containerSelector;
+};
+template <class Particle, class ParticleCell>
+template <class ParticleFunctor>
+void AutoTuner<Particle, ParticleCell>::iteratePairwise(ParticleFunctor *f, bool useSoA) {
+  bool newton3Allowed = f->allowsNewton3();
+  bool nonNewton3Allowed = f->allowsNonNewton3();
+  bool useNewton3 = false;
+  if (newton3Allowed and nonNewton3Allowed) {
+    /// @todo auto-tune (far off future)
+  } else if (not newton3Allowed and not nonNewton3Allowed) {
+    /// @todo throw exception
+  } else {
+    useNewton3 = newton3Allowed;
+  }
+
+  auto container = _containerSelector.getOptimalContainer();
+
+  /// @todo: WANT one single iteratePairwise(CellFunctor) for containers
+  /// @todo: CellFunctor for iteration should be build here using selectors for SoA and N3
+  if (useSoA) {
+    if (useNewton3) {
+      if (_iterationsSinceTuning >= _tuningInterval) {
+        tune<ParticleFunctor, true, true>(*f);
+      }
+      WithStaticContainerType(container, container->iteratePairwiseSoA(f, useNewton3););
```

Comment:
move these statements behind all the ifs, to deduplicate code a bit

Reply:
done
20847cfa204630cce4d33a37ef0d3a6788bdd939